codejail/jail.py

```diff
@@ -0,0 +1,329 @@
+        named in the `argv` list.
+
+        `files` is a list of file paths, they are all copied to the jailed
+        directory.  Note that no check is made here that the files don't contain
+        sensitive information.  The caller must somehow determine whether to allow
+        the code access to the files.  Symlinks will be copied as symlinks.  If the
+        linked-to file is not accessible to the sandbox, the symlink will be
+        unreadable as well.
+
+        `extra_files` is a list of pairs, each pair is a filename and a bytestring
+        of contents to write into that file.  These files will be created in the
+        temp directory and cleaned up automatically.  No subdirectories are
+        supported in the filename.
+
+        `argv` is the command-line arguments to supply.
+
+        `stdin` is a string, the data to provide as the stdin for the process.
+
+        `slug` is an arbitrary string, a description that's meaningful to the
+        caller, that will be used in log messages.
+
+        Return a JailResult namedtuple with:
+
+            .status: exit status of the process: an int, 0 for success
+            .stdout: stdout of the program, a string
+            .stderr: stderr of the program, a string
+
+        """
+        # We make a temp directory to serve as the home of the sandboxed code.
+        # It has a writable "tmp" directory within it for temp files.
+        with temp_directory() as homedir:
+
+            # Make directory readable by other users ('sandbox' user needs to be
+            # able to read it).
+            os.chmod(homedir, 0775)
+
+            # Make a subdir to use for temp files, world-writable so that the
+            # sandbox user can write to it.
+            tmptmp = os.path.join(homedir, "tmp")
+            os.mkdir(tmptmp)
+            os.chmod(tmptmp, 0777)
+
+            argv = argv or []
+            env = {'TMPDIR': 'tmp'}
+
+            # All the supporting files are copied into our directory.
+            for filename in files or ():
+                dest = os.path.join(homedir, os.path.basename(filename))
+                if os.path.islink(filename):
+                    os.symlink(os.readlink(filename), dest)
+                elif os.path.isfile(filename):
+                    shutil.copy(filename, homedir)
+                else:
+                    shutil.copytree(filename, dest, symlinks=True)
+
+            # Create the main file.
+            if code:
+                with open(os.path.join(homedir, "jailed_code"), "wb") as jailed:
+                    jailed.write(code)
+
+                argv = ["jailed_code"] + argv
+
+            # Create extra files requested by the caller:
+            for name, content in extra_files or ():
+                with open(os.path.join(homedir, name), "wb") as extra:
+                    extra.write(content)
+
+            cmd = []
+
+            # Build the command to run.
+            if self.user:
```

Comment:
I guess I meant is it desirable to allow it to be unspecified?  Is there a usecase that requires it?